Make subcell projection be dim-by-dim operation

nilsdeppe opened the pull request:
## Proposed changes

This speeds up subcell evolutions of GRMHD smooth flow by ~3x.

I'm going to add the option to do dim-by-dim reconstruction and to only run the TCI for going back to DG every `N` steps, but that'll be in a separate PR. I'm hoping to bring the overhead of DG-subcell down to < 5% when the solution is smooth with some more refactoring and changing order-of-operations in the code. That'll be a bit bigger of a refactor but I think will be mostly transparent from the user perspective. At that point we will need to profile the code again and make sure all of our time is spent in the RHS 🤞 

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Make subcell projection be dim-by-dim operation

This speeds up subcell evolutions of GRMHD smooth flow by ~3x

wthrowe commented on src/Evolution/DgSubcell/Matrices.cpp:
```diff
@@ -20,11 +20,76 @@
 template <Spectral::Quadrature QuadratureType, size_t NumDgGridPoints1d,
           size_t Dim>
-Matrix projection_matrix_cache_impl_helper(const Index<Dim>& subcell_extents) {
+Matrix projection_matrix_3d(const Index<Dim>& subcell_extents) {
```
Is this actually specific to 3D?  I don't see anything choosing functions based on dimension.

nilsdeppe replied:
ah, no. It's the "full dimensions". Do you have a good suggestion for the name? I was clearly focusing on the 3d case when I made the name change... I have a similar naming dilemma  for reconstruction. There's `DimByDim` but then for the full one do I do `AllDimensions`? Any suggestions are much appreciated!

nilsvu commented on src/Evolution/DgSubcell/Matrices.cpp:
```diff
@@ -20,11 +20,76 @@
+              Spectral::minimum_number_of_points<
+                  Spectral::Basis::Legendre,
+                  Spectral::Quadrature::GaussLobatto>,
+              Spectral::maximum_number_of_points<Spectral::Basis::Legendre>>,
```
The `End` of a `CacheRange` is exclusive. Do you want to add a `+1` here (and below)? If no, please add a comment.

nilsvu commented on src/Evolution/DgSubcell/Matrices.cpp:
```diff
@@ -20,11 +20,76 @@
 namespace evolution::dg::subcell::fd {
+template <size_t Dim>
+const Matrix& projection_matrix(const Mesh<Dim>& dg_mesh,
+                                const Index<Dim>& subcell_extents) {
```
Since this function returns a 1D matrix, how about you take a `Mesh<1> dg_mesh` and a `size_t subcell_extents` here? This is how it's done in `NumericalAlgorithms/Spectral/Projection.hpp`. There's also an overload that returns the 3D matrix (as a `std::array<std::reference_wrapper<const Matrix>, Dim>`), taking a `Mesh<Dim>` and `Index<Dim>`.

kidder reviewed (commented):
I have no requested changes

nilsdeppe commented:
Rebased and squashed in changes since the number of changes for the fixup was close to half the changes of the original commit :) Thanks for the reviews!

nilsvu reviewed (approved):
LGTM
